src/main.rs

```diff
@@ -128,6 +150,155 @@ fn gen(cfg: &Config, markdown: String, out: impl Write) {
     }
 }
 
+fn ensure_mp4_tools_installed() {
+    // TODO: minimum version requirements or alternatives?
+    let has_pdf_to_ppm = Command::new("pdftoppm")
+        .arg("-v")
+        .status()
+        .map(|status| status.success())
+        .unwrap_or(false);
+
+    if !has_pdf_to_ppm {
+        panic!("The tool `pdftoppm` is required but does not appear to be installed (it's a part of poppler).");
+    }
+
+    // TODO: minimum version requirements?
+    let has_ffmpeg = Command::new("ffmpeg")
+        .arg("-version")
+        .status()
+        .map(|status| status.success())
+        .unwrap_or(false);
+
+    if !has_ffmpeg {
+        panic!("The tool `ffmpeg` is required but does not appear to be installed.");
+    }
+
+    let has_ffprobe = Command::new("ffprobe")
+        .arg("-version")
+        .status()
+        .map(|status| status.success())
+        .unwrap_or(false);
+
+    if !has_ffprobe {
+        panic!("The tool `ffprobe` is required but does not appear to be installed (it's usually a part of ffmpeg).");
+    }
+
+    // TODO: minimum version requirements and alternatives?
+    let has_espeak_ng = Command::new("espeak-ng")
+        .arg("--version")
+        .status()
+        .map(|status| status.success())
+        .unwrap_or(false);
+
+    if !has_espeak_ng {
+        panic!("The tool `espeak-ng` is required but does not appear to be installed.");
+    }
+}
+
+fn ffmpeg<P: AsRef<Path>>(pdf: P, cfg: &Config) -> PathBuf {
+    // Start by demuxing the pdf into its frames, which gives us their number.
+    // We use poppler tools and not imagemagick because the latter is bloody stupid. It
+    // disables code access to pdf conversion as a 'security measure'. That's an idiots patch
+    // to broken code that just hurts usability. There's nothing malicious about what we're
+    // attempting to do but, I assumed, since they can't admit to the PHP crowds of SaaS image
+    // tools that their tools are utterly messy and completely unmaintainable they introduce a
+    // black-list that makes it impossible to use for other people as well.
+    // /rant
+    Command::new("pdftoppm")
+        .current_dir(&cfg.out_dir)
+        .arg("-png")
```

Comment:
For better Resolution of the images in the mp4 (currently they only have 100 or 150 dpi, which looks pixelated).

```suggestion
        .args(&["-png", "-rx", "600", "-ry", "600"])
```